### DISCUSSION TITLE
Possible integration with HoloViews/Datashader for unstructured grid (MPAS) visualization.

### DISCUSSION THREAD

**Discussion by cdeciampa** [UXARRAY/uxarray#105] [Ideas]
For the past month or so, I've been using [Datashader](https://datashader.org/index.html) to visualize unstructured grid outputs from the MPAS model without the need for regridded MPAS files. A sample notebook using Datashader and the MPAS model [can be viewed here](https://geocat-examples.readthedocs.io/en/latest/gallery-notebooks/Datashader/MPAS_Datashader_Trimesh.html). My code is very similar to the sample notebook and uses the same naming conventions for consistency.

Datashader uses HoloViews (which piggybacks off of Bokeh or Matplotlib) to output interactive global plots, which is ideal for analysis within Jupyter notebooks. Admittedly, the syntax for plotting parameter options (title, colorbar, fonts, markers, etc.) is clunky and not intuitive and the docs aren't much help when it comes to specifics, so a wrapper function in UXARRAY would be great.

```
rasterized.opts(tools=['hover'], colorbar=True, cmap='coolwarm') *\
gf.coastline(projection=projection)
```

Similarly, Datashader resamples the data through the `rasterize` function (imported below as `hds_rasterize`) by default, which would also benefit from a wrapper function as its parameters are a bit unclear as well.

```
rasterized = hds_rasterize(trimesh, aggregator='mean', precompute=True, 
height=pixel_height, width=pixel_width, pixel_ratio=pixel_ratio,
x_sampling=x_sampling, y_sampling=y_sampling)
```

Both of the images below are 3 km resolution of an artificial tropical cyclone, but Image 1 has `pixel_ratio=1` and Image 2 has `pixel_ratio=10`.

![Image 1](https://user-images.githubusercontent.com/66640080/184672168-8d1945fc-2402-4062-a05b-481a2953b14c.png)
![Image 2](https://user-images.githubusercontent.com/66640080/184672163-4e84783f-5806-4a97-bfeb-3b0b29e8a85c.png)

All that to say, I think Datashader has the potential for interactive visualization for native MPAS output and could be integrated into UXARRAY.

**Comment by erogluorhan** [comment]
Hi @cdeciampa , thanks very much for your input! Just an FYI, we gave out that sample notebook together as the GeoCAT team, and we have been further exploring the Datashader stack for use in the unstructured gross visualization. We just wrapped up a winner internship project on that, too, and we will keep investigating how to utilize it for Uxarray, maybe integrate through lightweight, simplified interfaces as you pointed out.

Stay tuned and keep sharing your feedback with us as well as feel free to contribute in any direction!

**Reply by paullric** [reply]
As part of a related project I've been developing a ncview alternative that works directly on unstructured grid data (https://github.com/SEATStandards/ncvis).  While it doesn't provide any python bindings at this point, you may be interested in playing around with it for "quick look" visualizations.  With the kd-tree based grid sampler it is able to capture the centroidal voronoi grid cells from the MPAS grid.

<img width="842" alt="ncvis_tc_screenshot" src="https://user-images.githubusercontent.com/5330916/185259901-edc7a788-de65-4a8f-a50a-75d8efa91c85.png">

<img width="842" alt="ncvis_tc_screenshot2" src="https://user-images.githubusercontent.com/5330916/185260177-53dcd9de-2785-490f-86bd-b82bdf460521.png">

**Reply by zarzycki** [reply]
@paullric, this is promising that ncvis can capture the underlying grid structure with the kd-sampling. I assume the current thought for ncvis is to remain essentially the successor to ncview, which is excellent for quick/dirty evaluation but probably not suitable for publication quality graphics (e.g., like how I routinely use ncview up until it's time to package things for viz to non-modelers, at which point I use NCL, etc.).

**Reply by paullric** [reply]
Yes, that's the plan at this point.  Of course, that depends on how high you set the bar for "publication quality."

**Reply by zarzycki** [reply]
True. And to be fair, if you are skilled, you can make ncview image dumps that supersede the quality of some things I review...

Mostly thinking about how we want uxarray to fit into the viz ecosystem w/r/t unstructured grids. I assume the path of least resistance is to provide objects describe cell values and information about the mesh topology (e.g., nsides, grid edges, etc.) and let dedicated viz tools rasterize that, but that information is only good if a viz tool like cartopy, datashader, etc. can easily use it...

**Comment by philipc2** [comment]
Hi @cdeciampa 

Thanks for the suggestion. We've had a short discussion about this in #118. We are very much still coming up with the design for these wrappers, so any suggestions are more than appriciated! 

Have you looked into using hvPlot paired with GeoViews? I worked on researching unstructured grid visualization this summer, and hvPlot acts as a high-level API for Datashader. I used hvPlot and GeoViews in my example notebooks from my summer project which can be found here https://github.com/NCAR/geocat-scratch/tree/main/polymesh

**Reply by cdeciampa** [reply]
Thank you for steering me in this direction! I finally had the time to dig into your source code and modified it for my needs with MPAS (since it's not currently supported by uxarray), but I was able to plot the hexagonal grid thanks to your really great work.

![bokeh_plot-41](https://user-images.githubusercontent.com/66640080/213896502-a7c44e7b-c8c8-4373-9a83-955f847d2029.png)

**Reply by philipc2** [reply]
Hi @cdeciampa 

This is fantastic! I'm glad that I was able to point you in the right direction and the visualizations look great. I'll give this a deeper look when I have some time.

We are currently working to resolve #175 and are about to have a team-wide discussion on the design behind our visualization implementation this week. In the end you should expect UXarray to be able to directly read in MPAS files and support visualizations similar to this one directly with accessors/wrappers. 

Expect a new discussion post sometime next week regarding our next steps with visualization!
